2_process/src/process_nhdv2_attr.R

```diff
@@ -269,3 +269,184 @@ process_catchment_nhdv2_attr <- function(file_path,vars_table,segs_w_comids,nhd_
+    TAV_segs <- filter(nhdv2_attr_refined, hru_segment %in% ind_segs) %>% 
+      pull(CAT_TAV7100_ANN_area_wtd)
+    TMN_segs <- filter(nhdv2_attr_refined, hru_segment %in% ind_segs) %>% 
+      pull(CAT_TMIN7100_area_wtd)
+    STR_DENS_segs <- filter(nhdv2_attr_refined, hru_segment %in% ind_segs) %>% 
+      pull(CAT_STRM_DENS_area_wtd)
+    
+    #Compute TOT values
+    #need special handling for PRMS segments with _2
+    ind_replace <- which(nhdv2_attr_refined$hru_segment == subseg)
+    nhdv2_attr_refined$TOT_CWD_frmCAT[ind_replace] <- compute_tot(CWD_segs,areas_segs)
+    nhdv2_attr_refined$TOT_TAV7100_ANN_frmCAT[ind_replace] <- compute_tot(TAV_segs,areas_segs)
+    nhdv2_attr_refined$TOT_TMIN7100_frmCAT[ind_replace] <- compute_tot(TMN_segs,areas_segs)
+    nhdv2_attr_refined$TOT_STRM_DENS_frmCAT[ind_replace] <- compute_tot(STR_DENS_segs,areas_segs)
+  }
+  
+  return(nhdv2_attr_refined)
+}
+
+refine_from_neighbors <- function(nhdv2_attr, attr_i, prms_reach_attr
+                                  ){
+  #' 
+  #' @description Function to fill in a reach's attribute value with a value 
+  #' from its neighboring reaches (from_segs and to_seg)
+  #'
+  #' @param nhdv2_attr the full (or partially refined) attribute tbl
+  #' @param attr_i the attribute name as character string
+  #' @param prms_reach_attr the PRMS reach attribute tbl
+  #'
+  #' @value Returns nhdv2_attr with filled in values for the attr_i column.
+  
+  if(attr_i == 'CAT_EWT_area_wtd'){
+    #Search for < -100
+    ind_reach <- filter(nhdv2_attr, get(attr_i) < -100) %>%
+      pull(PRMS_segid)
+  }else{
+    #search for NAs
+    ind_reach <- filter(nhdv2_attr, is.na(get(attr_i))) %>%
+      pull(PRMS_segid)
+  }
+  #find the from and to segments for this reach
+  seg_match <- filter(prms_reach_attr, subseg_id == ind_reach) %>%
+    select(from_segs, to_seg) %>%
+    mutate(from_segs = str_split(from_segs, pattern = ';', simplify = F)) %>%
+    mutate(segs = list(c(from_segs[[1]], to_seg))) %>%
+    select(-from_segs, -to_seg) %>%
+    unlist() %>%
+    #add _1 to match PRMS seg ID
+    paste0(., '_1')
+  #get the average of the attributes for the matched reaches
+  fill_val <- filter(nhdv2_attr, PRMS_segid %in% seg_match) %>%
+    select(all_of(attr_i)) %>%
+    colMeans() %>%
+    as.numeric()
+  #assign to attribute table
+  nhdv2_attr_refined <- mutate(nhdv2_attr,
+                               attr = case_when(PRMS_segid == ind_reach ~ fill_val,
+                                                              TRUE ~ get(attr_i))
+                               ) %>%
+    pull(attr)
+  return(nhdv2_attr_refined)
+}
+
+compute_tot <- function(cat_var, areas_segs){
+  #' 
+  #' @description Function to compute TOT from CAT variables
+  #'
+  #' @param cat_var vector of CAT variables for current and upstream segments
+  #' @param areas_segs area of each segment in the same order as cat_vec
+  #'
+  #' @value Returns the area-weighted sum (TOT) variable
```

Comment:
isn't this an area-weighted mean, with area_segs as the weight.

Reply:
yes, thanks for catching the typo!